### PR TITLE
Remove some uses of `guess_head_span`

### PR DESCRIPTION
**Pull request by cjgillot**
That function cuts a span at the first occurrence of `{`.  Using `def_span` is almost always more precise.

### CODE REVIEW

File: src/test/ui/cycle-trait/cycle-trait-default-type-trait.stderr [hunk 1]
```diff
@@ -8,8 +8,13 @@ LL | trait Foo<X = Box<dyn Foo>> {
 note: cycle used when collecting item types in top-level module
   --> $DIR/cycle-trait-default-type-trait.rs:4:1
    |
-LL | trait Foo<X = Box<dyn Foo>> {
-   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+LL | / trait Foo<X = Box<dyn Foo>> {
```
Comment (cjgillot) [inline src/test/ui/cycle-trait/cycle-trait-default-type-trait.stderr:11]: This span was wrong, as it's supposed to point to the module.

File: compiler/rustc_query_system/src/query/job.rs [hunk 2]
```diff
@@ -536,9 +535,7 @@ pub(crate) fn report_cycle<'a>(
 ) -> DiagnosticBuilder<'a, ErrorGuaranteed> {
     assert!(!stack.is_empty());
 
-    let fix_span = |span: Span, query: &QueryStackFrame| {
-        sess.source_map().guess_head_span(query.default_span(span))
-    };
+    let fix_span = |span: Span, query: &QueryStackFrame| query.default_span(span);
```
Comment (davidtwco) [inline compiler/rustc_query_system/src/query/job.rs:538]: nit: does it still make sense to have a closure for this?

**Review by davidtwco — APPROVED** [review]
One nit, otherwise LGTM, r=me when CI passes

**Review by davidtwco — APPROVED** [review]

**Pull request merged**